### PR TITLE
8295270: RISC-V: Clean up and refactoring for assembler functions

### PR DESCRIPTION
**Pull request by zifeihan**
Please review this backport to riscv-port-jdk17u.
Backport of [JDK-8295270](https://bugs.openjdk.org/browse/JDK-8295270).
[JDK-8295270](https://bugs.openjdk.org/browse/JDK-8295270) not clean，origin patch has no eden_allocate, which removed with [JDK-8290706](https://bugs.openjdk.org/browse/JDK-8290706)

Testing:

Tier1-3 passed without new failure on unmacthed (release).

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8295270](https://bugs.openjdk.org/browse/JDK-8295270): RISC-V: Clean up and refactoring for assembler functions

### Reviewers
 * [Xiaolin Zheng](https://openjdk.org/census#xlinzheng) (@zhengxiaolinX - Author)
 * [Fei Yang](https://openjdk.org/census#fyang) (@RealFYang - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/riscv-port-jdk17u.git pull/38/head:pull/38` \
`$ git checkout pull/38`

Update a local copy of the PR: \
`$ git checkout pull/38` \
`$ git pull https://git.openjdk.org/riscv-port-jdk17u.git pull/38/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 38`

View PR using the GUI difftool: \
`$ git pr show -t 38`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/riscv-port-jdk17u/pull/38.diff">https://git.openjdk.org/riscv-port-jdk17u/pull/38.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/riscv-port-jdk17u/pull/38#issuecomment-1500782483)

### CODE REVIEW

File: src/hotspot/cpu/riscv/assembler_riscv.hpp [hunk 1]
```diff
@@ -488,14 +449,7 @@ class Assembler : public AbstractAssembler {
 
 #undef INSN
 
-#define INSN_ENTRY_RELOC(result_type, header)                               \
-  result_type header {                                                      \
-    guarantee(rtype == relocInfo::internal_word_type,                       \
-              "only internal_word_type relocs make sense here");            \
-    relocate(InternalAddress(dest).rspec());                                \
-    IncompressibleRegion ir(this);  /* relocations */
```
Comment (zhengxiaolinX) [inline src/hotspot/cpu/riscv/assembler_riscv.hpp:496]: Here we have an `IncompressibleRegion` thing to keep instructions in `INSN_ENTRY_RELOC` from being compressed by the C extension, for relocations are here. But after moving to different locations in this patch, the `IncompressibleRegion` is missing. Please see the below comments.

File: src/hotspot/cpu/riscv/macroAssembler_riscv.hpp [hunk 2]
```diff
@@ -443,28 +442,83 @@ class MacroAssembler: public Assembler {
...
+  void bgt (Register Rs, Register Rt, Label &l, bool is_far = false);
+  void ble (Register Rs, Register Rt, Label &l, bool is_far = false);
+  void bgtu(Register Rs, Register Rt, Label &l, bool is_far = false);
+  void bleu(Register Rs, Register Rt, Label &l, bool is_far = false);
+
+#define INSN_ENTRY_RELOC(result_type, header)                               \
+  result_type header {                                                      \
+    guarantee(rtype == relocInfo::internal_word_type,                       \
+              "only internal_word_type relocs make sense here");            \
+    relocate(InternalAddress(dest).rspec());
```
Comment (zhengxiaolinX) [inline src/hotspot/cpu/riscv/macroAssembler_riscv.hpp:491]:
```suggestion
    relocate(InternalAddress(dest).rspec());                  \
    IncompressibleRegion ir(this); /* relocations */
```

File: src/hotspot/cpu/riscv/macroAssembler_riscv.hpp [hunk 3]
```diff
@@ -547,6 +645,215 @@ class MacroAssembler: public Assembler {
   void andi(Register Rd, Register Rn, int64_t imm, Register tmp = t0);
   void orptr(Address adr, RegisterOrConstant src, Register tmp1 = t0, Register tmp2 = t1);
 
+// Load and Store Instructions
+#define INSN_ENTRY_RELOC(result_type, header)                               \
+  result_type header {                                                      \
+    guarantee(rtype == relocInfo::internal_word_type,                       \
+              "only internal_word_type relocs make sense here");            \
+    relocate(InternalAddress(dest).rspec());
```
Comment (zhengxiaolinX) [inline src/hotspot/cpu/riscv/macroAssembler_riscv.hpp:653]:
```suggestion
    relocate(InternalAddress(dest).rspec());                  \
    IncompressibleRegion ir(this); /* relocations */
```

File: src/hotspot/cpu/riscv/assembler_riscv.hpp (hunk 1 above)
Comment (zifeihan) [reply src/hotspot/cpu/riscv/assembler_riscv.hpp:496]: Thanks for the reivew, fixed.

**Review by zhengxiaolinX — APPROVED** [review]
Thanks for these incremental changes! LGTM from a cursory look.

**Review by RealFYang — APPROVED** [review]
Looks good. Thanks.

**Comment by zifeihan** [comment]
Thanks all for the review.
/integrate

**Pull request closed without merging**